### ISSUE TITLE
zsh: /opt/homebrew/bin/pnpm: bad interpreter: /opt/homebrew/opt/node/bin/node: no such file or directory

### ISSUE THREAD

**Issue by gfazioli** [pnpm/pnpm#8384]
### Verify latest release

- [X] I verified that the issue exists in the latest pnpm release

### pnpm version

9.7.0

### Which area(s) of pnpm are affected? (leave empty if unsure)

Operating System (Windows, MacOS, Linux)

### Reproduction steps

stop working after an usual
```sh
brew update && brew upgrade && brew doctor
````

### Describe the Bug

```sh
> pnpm -v
zsh: /opt/homebrew/bin/pnpm: bad interpreter: /opt/homebrew/opt/node/bin/node: no such file or directory
```
I had to remove `pnpm` from brew and install it globally using `npm`.

### Expected Behavior

```sh
> pnpm -v
9.7.0
```

### Which Node.js version are you using?

v22.5.1

### Which operating systems have you used?

- [X] macOS
- [ ] Windows
- [ ] Linux

**Comment by gfazioli** [comment]
By using
```sh
brew install --build-from-source pnpm
```

seems work

**Comment by aholland** [comment]
My solution was to edit `/opt/homebrew/bin/pnpm` ie `/opt/homebrew/bin/pnpm@ -> ../Cellar/pnpm/9.7.0/bin/pnpm`,
changing the first line to use node from my path instead of assuming I have node installed via Homebrew. I have nvm installed via Homebrew and node installed via nvm. The unedited version of the first line is shown on the second line, commented out.

```
#!/usr/bin/env node
// #!/opt/homebrew/opt/node/bin/node
const [major, minor] = process.version.slice(1).split('.')
const COMPATIBILITY_PAGE = `Visit https://r.pnpm.io/comp to see the list of past pnpm versions with respective Node.js version support.`

// We don't use the semver library here because:
//  1. it is already bundled to dist/pnpm.cjs, so we would load it twice
//  2. we want this file to support potentially older Node.js versions than what semver supports
if (major < 18 || major == 18 && minor < 12) {
  console.log(`ERROR: This version of pnpm requires at least Node.js v18.12
The current version of Node.js is ${process.version}
${COMPATIBILITY_PAGE}`)
  process.exit(1)
}

// We need to load v8-compile-cache.js separately in order to have effect
try {
  require('v8-compile-cache');
} catch {
  // We don't have/need to care about v8-compile-cache failed
}

global['pnpm__startedAt'] = Date.now()
require('../dist/pnpm.cjs')

// if you want to debug at your local env, you can use this
// require('../lib/pnpm')

```

**Comment by zkochan** [comment]
Is this something that homebrew has changed recently? I don't think we did anything that could cause such issues in 9.7

**Comment by gfazioli** [comment]
@zkochan to be honest, I don't know. I launch every morning 

```sh
brew update && brew upgrade && brew doctor
```

and yesterday it stopped working... both on MacOS Sonoma and on Monterey

**Comment by birdzai** [comment]
Something clearly broke with pnpm 9.7.0 and homebrew-installed node. Everything worked up until yesterday when the new pnpm was updated via homebrew.

**Comment by zkochan** [comment]
is it still broken if you downgrade to pnpm 9.6?

**Comment by birdzai** [comment]
> is it still broken if you downgrade to pnpm 9.6?

Can't downgrade to 9.6 using homebrew (can only downgrade to pnpm v8.x, as far as I can tell....correct me if I'm wrong). I would simply like to have it working as it always did with brew-installed node. it was working 2 days ago before upgrading to 9.7

**Comment by zkochan** [comment]
> My solution was to edit /opt/homebrew/bin/pnpm ie /opt/homebrew/bin/pnpm@ -> ../Cellar/pnpm/9.7.0/bin/pnpm,
changing the first line to use node from my path instead of assuming I have node installed via Homebrew. I have nvm installed via Homebrew and node installed via nvm. The unedited version of the first line is shown on the second line, commented out.

If you check our source code, we use node.js from the PATH:

https://github.com/pnpm/pnpm/blob/7fa4599dbc8a92383e39a3185df540048f62e4f9/pnpm/bin/pnpm.cjs#L1 

Something is changing that shebang. We don't change it.

**Comment by joshswan** [comment]
@zkochan Looks like it was a change in Homebrew: https://github.com/Homebrew/homebrew-core/issues/176257

**Comment by birdzai** [comment]
I don't think modifying the brew-installed pnpm script is a solution as you'd be having to do this after every pnpm formula update. 

For nvm users, would uninstalling the homebrew node installation resolve this?

**Comment by Hmoulvad** [comment]
Have anyone found a fix yet to this issue. I just updated today, and got the same error.

**Comment by gfazioli** [comment]
> Have anyone found a fix yet to this issue. I just updated today, and got the same error.

certainly the issue seems to involve those who use `nvm`
I solved it by uninstalling `pnpm` with brew and then reinstalling it with

```sh
brew install --build-from-source pnpm
```

**Comment by kjxbyz** [comment]
same issue, using nvm.

**Comment by kjxbyz** [comment]
Install Older Versions of Homebrew Packages:

1. `brew uninstall pnpm`
2. Download [pnpm.rb](https://github.com/Homebrew/homebrew-core/blob/c3009c1507f38fae277faa04500e64b357c72c84/Formula/p/pnpm.rb)
3. `pnpm install pnpm.rb`

**Comment by SenseiMarv** [comment]
The Homebrew team has looked into this, and this should fix the issue when it's released: https://github.com/Homebrew/brew/pull/17993

**Comment by michug** [comment]
> My solution was to edit `/opt/homebrew/bin/pnpm` ie `/opt/homebrew/bin/pnpm@ -> ../Cellar/pnpm/9.7.0/bin/pnpm`, changing the first line to use node from my path instead of assuming I have node installed via Homebrew. I have nvm installed via Homebrew and node installed via nvm. The unedited version of the first line is shown on the second line, commented out.
> 
> ```
> #!/usr/bin/env node
> // #!/opt/homebrew/opt/node/bin/node
> const [major, minor] = process.version.slice(1).split('.')
> const COMPATIBILITY_PAGE = `Visit https://r.pnpm.io/comp to see the list of past pnpm versions with respective Node.js version support.`
> 
> // We don't use the semver library here because:
> //  1. it is already bundled to dist/pnpm.cjs, so we would load it twice
> //  2. we want this file to support potentially older Node.js versions than what semver supports
> if (major < 18 || major == 18 && minor < 12) {
>   console.log(`ERROR: This version of pnpm requires at least Node.js v18.12
> The current version of Node.js is ${process.version}
> ${COMPATIBILITY_PAGE}`)
>   process.exit(1)
> }
> 
> // We need to load v8-compile-cache.js separately in order to have effect
> try {
>   require('v8-compile-cache');
> } catch {
>   // We don't have/need to care about v8-compile-cache failed
> }
> 
> global['pnpm__startedAt'] = Date.now()
> require('../dist/pnpm.cjs')
> 
> // if you want to debug at your local env, you can use this
> // require('../lib/pnpm')
> ```

This is the correct explanation of the error message. The solution worked perfectly, just wanted to add that I had to replace the first line with another path since I had node installed in another folder, to know where is yours run ```which npm```

**Comment by birdzai** [comment]
> > My solution was to edit `/opt/homebrew/bin/pnpm` ie `/opt/homebrew/bin/pnpm@ -> ../Cellar/pnpm/9.7.0/bin/pnpm`, changing the first line to use node from my path instead of assuming I have node installed via Homebrew. I have nvm installed via Homebrew and node installed via nvm. The unedited version of the first line is shown on the second line, commented out.
> > ```
> > #!/usr/bin/env node
> > // #!/opt/homebrew/opt/node/bin/node
> > const [major, minor] = process.version.slice(1).split('.')
> > const COMPATIBILITY_PAGE = `Visit https://r.pnpm.io/comp to see the list of past pnpm versions with respective Node.js version support.`
> > 
> > // We don't use the semver library here because:
> > //  1. it is already bundled to dist/pnpm.cjs, so we would load it twice
> > //  2. we want this file to support potentially older Node.js versions than what semver supports
> > if (major < 18 || major == 18 && minor < 12) {
> >   console.log(`ERROR: This version of pnpm requires at least Node.js v18.12
> > The current version of Node.js is ${process.version}
> > ${COMPATIBILITY_PAGE}`)
> >   process.exit(1)
> > }
> > 
> > // We need to load v8-compile-cache.js separately in order to have effect
> > try {
> >   require('v8-compile-cache');
> > } catch {
> >   // We don't have/need to care about v8-compile-cache failed
> > }
> > 
> > global['pnpm__startedAt'] = Date.now()
> > require('../dist/pnpm.cjs')
> > 
> > // if you want to debug at your local env, you can use this
> > // require('../lib/pnpm')
> > ```
> 
> This is the correct explanation of the error message. The solution worked perfectly, just wanted to add that I had to replace the first line with another path since I had node installed in another folder, to know where is yours run `which npm`

Your solution would only work temporarily until the next time Homebrew updated pnmp, in which your edited shebang line would get re-written. The Homebrew team has fixed the issue and the shebang line will no longer be overwritten for formula's that don't have a hard dependency on node. I'll wait for the updated pnpm to be released.

**Comment by ncheungg** [comment]
If you already have pnpm installed, you can run this to fix it
```
brew reinstall --build-from-source pnpm
```

**Comment by Nehasoni988** [comment]
These steps worked for me-

1. Uninstall the `pnpm`-

Either with-
```
npm rm -g pnpm
````
Or with-
```
rm -rf $PNPM_HOME
```
2. Re-install it using `curl` ([Reference link](https://pnpm.io/installation#on-posix-systems))
```
curl -fsSL https://get.pnpm.io/install.sh | sh -
```
Now, check the `pnpm` version-
```
pnpm -v
```

**Comment by kjxbyz** [comment]
[Homebrew v4.3.15](https://github.com/Homebrew/brew/releases/tag/4.3.15) has been released, but the issue still exists

**Comment by SenseiMarv** [comment]
> > The Homebrew team has looked into this, and this should fix the issue when it's released: [Homebrew/brew#17993](https://github.com/Homebrew/brew/pull/17993)
> 
> [Homebrew v4.3.15](https://github.com/Homebrew/brew/releases/tag/4.3.15) has been released, but the issue still exists

They also had to update the pnpm formula. This has been done [a few hours ago](https://github.com/Homebrew/homebrew-core/commit/a605de482e38f8dbcce629319d1f631d75749222). So please ensure that you are at least on Homebrew 4.3.15 and pnpm formula 9.7.0_1. I updated to those and can confirm that pnpm is now using the correct version again.

**Comment by kjxbyz** [comment]
It works for me. Thank you.
